libs/@local/hashql/hir/src/fold/mod.rs

```diff
@@ -0,0 +1,732 @@
+        variable: LocalVariable<'heap>,
+    ) -> Self::Output<LocalVariable<'heap>> {
+        walk_local_variable(self, variable)
+    }
+
+    fn fold_qualified_variable(
+        &mut self,
+        variable: QualifiedVariable<'heap>,
+    ) -> Self::Output<QualifiedVariable<'heap>> {
+        walk_qualified_variable(self, variable)
+    }
+
+    fn fold_let(&mut self, r#let: Let<'heap>) -> Self::Output<Let<'heap>> {
+        walk_let(self, r#let)
+    }
+
+    fn fold_input(&mut self, input: Input<'heap>) -> Self::Output<Input<'heap>> {
+        walk_input(self, input)
+    }
+
+    fn fold_operation(&mut self, operation: Operation<'heap>) -> Self::Output<Operation<'heap>> {
+        walk_operation(self, operation)
+    }
+
+    fn fold_type_operation(
+        &mut self,
+        operation: TypeOperation<'heap>,
+    ) -> Self::Output<TypeOperation<'heap>> {
+        walk_type_operation(self, operation)
+    }
+
+    fn fold_type_assertion(
+        &mut self,
+        assertion: TypeAssertion<'heap>,
+    ) -> Self::Output<TypeAssertion<'heap>> {
+        walk_type_assertion(self, assertion)
+    }
+
+    fn fold_type_constructor(
+        &mut self,
+        constructor: TypeConstructor<'heap>,
+    ) -> Self::Output<TypeConstructor<'heap>> {
+        walk_type_constructor(self, constructor)
+    }
+
+    fn fold_binary_operation(
+        &mut self,
+        operation: BinaryOperation<'heap>,
+    ) -> Self::Output<BinaryOperation<'heap>> {
+        walk_binary_operation(self, operation)
+    }
+
+    fn fold_unary_operation(
+        &mut self,
+        operation: UnaryOperation<'heap>,
+    ) -> Self::Output<UnaryOperation<'heap>> {
+        walk_unary_operation(self, operation)
+    }
+
+    fn fold_access(&mut self, access: Access<'heap>) -> Self::Output<Access<'heap>> {
+        walk_access(self, access)
+    }
+
+    fn fold_field_access(
+        &mut self,
+        access: FieldAccess<'heap>,
+    ) -> Self::Output<FieldAccess<'heap>> {
+        walk_field_access(self, access)
+    }
+
+    fn fold_index_access(
+        &mut self,
+        access: IndexAccess<'heap>,
+    ) -> Self::Output<IndexAccess<'heap>> {
+        walk_index_access(self, access)
+    }
+
+    fn fold_call(&mut self, call: Call<'heap>) -> Self::Output<Call<'heap>> {
+        walk_call(self, call)
+    }
+
+    fn fold_branch(&mut self, branch: Branch<'heap>) -> Self::Output<Branch<'heap>> {
+        walk_branch(self, branch)
+    }
+
+    fn fold_closure(&mut self, closure: Closure<'heap>) -> Self::Output<Closure<'heap>> {
+        walk_closure(self, closure)
+    }
+
+    fn fold_closure_signature(
+        &mut self,
+        signature: ClosureSignature<'heap>,
+    ) -> Self::Output<ClosureSignature<'heap>> {
+        walk_closure_signature(self, signature)
+    }
+
+    fn fold_graph(&mut self, graph: Graph<'heap>) -> Self::Output<Graph<'heap>> {
+        walk_graph(self, graph)
+    }
+}
+
+pub fn walk_ident<'heap, T: Fold<'heap> + ?Sized>(
```

Comment:
Is `Fold` dyn-safe?

Reply:
Fold is not dyn safe (and doesn't need to be - at least for the foreseeable future ) because of the GAT.

Reply:
In theory we don't but I copy-pasted the visitor code so that was left over. As you've pointed out it doesn't really hurt and makes migration in the future easier I guess^^.